File: ml-a-z/Polynomial_Regression/milam_poly.py
# Polynomial Regression
# %%
# import the libraries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder, StandardScaler, PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from pprint import pprint

%matplotlib inline

# suppress printing of scientific notation
np.set_printoptions(suppress=True)

# %%
# import the dataset
dataset = pd.read_csv('Position_Salaries.csv')

# %%
# create features matrix
X = dataset.iloc[:, 1:-1].values
# create dependent variable vector
y = dataset.iloc[:, -1].values

# %%
# fit linear regression to the dataset
linear_regr = LinearRegression()
linear_regr.fit(X, y)

# %%
# fit polynomial regression to the dataset
poly_regr = PolynomialFeatures(4)
X_poly = poly_regr.fit_transform(X)

# %%
# fit X_poly to new linear regression
linear_regr2 = LinearRegression()
linear_regr2.fit(X_poly, y)

# %%
# visualize the results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X, linear_regr.predict(X), color = 'blue')
plt.plot(X_grid, linear_regr2.predict(poly_regr.fit_transform(X_grid)), color = 'steelblue')
plt.title('Truth vs Bluff')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# %%
# predict new result with linear regression and polynomial regression
linear_regr.predict(6.5)
linear_regr2.predict(poly_regr.fit_transform(6.5))

File: ml-a-z/Polynomial_Regression/milam_poly.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# fit linear regression to dataset
lin_regr = lm(formula = Salary ~ ., data = dataset)


# fit polynomial regression to dataset
dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4
poly_regr = lm(formula = Salary ~ ., data = dataset)

# visualize the results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), 
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_regr, newdata = dataset)), color = 'blue') +
  geom_line(aes(x = dataset$Level, y = predict(poly_regr, newdata = dataset)), color = 'steelblue') +
  ggtitle('Truth or Bluff') +
  xlab('Level') +
  ylab('Salary')

# predict a new single result
new_pred = predict(lin_regr, data.frame(Level = 6.5))
new_poly_pred = predict(poly_regr, data.frame(Level = 6.5, 
                                              Level2 = 6.5^2, 
                                              Level3 = 6.5^3, 
                                              Level4 = 6.5^4))
